pkg/admission/pod_eviction.go

```diff
@@ -0,0 +1,254 @@
+		return request.allowWithReason("unable to find label on pod")
+	}
+
+	// ie ingester
+	request.log.SetSpanAndLogTag(config.RolloutGroupLabelKey, pod.Labels[config.RolloutGroupLabelKey])
+
+	// get the StatefulSet who manages this pod
+	if sts, err = request.clients.owner(pod); err != nil {
+		_ = level.Error(request.log).Log(logMsg, fmt.Sprintf("%s - unable to find pod owner", logAllow), logErr, err)
+		return request.allowWithReason("unable to find a StatefulSet as pod owner")
+	}
+
+	// ie ingester-zone-a
+	request.log.SetSpanAndLogTag("owner", sts.Name)
+
+	PdbConfig, err := config.GetCustomResourceConfig(request.clients.ctx, request.req.Request.Namespace, pod.Labels[config.RolloutGroupLabelKey], request.clients.dynamicClient, request.log)
+	if err != nil {
+		return request.allowWithReason(err.Error())
+	}
+
+	// the number of allowed unavailable pods in other zones.
+	maxUnavailable := PdbConfig.MaxUnavailablePods(sts)
+	if maxUnavailable == 0 {
+		_ = level.Info(request.log).Log(logMsg, fmt.Sprintf("%s - %s", logDeny, logUnavailable0))
+		return request.denyWithReason(logUnavailable0)
+	}
+
+	// Find the other StatefulSets which span all zones.
+	// Assumption - each StatefulSet manages all the pods for a single zone. This list of StatefulSets covers all zones.
+	// During a migration it may be possible to break this assumption, but during a migration the maxUnavailable will be set to 0 and the eviction request will be denied.
+	otherStatefulSets, err := request.clients.findRelatedStatefulSets(sts, PdbConfig.StsSelector())
+	if err != nil || otherStatefulSets == nil || len(otherStatefulSets.Items) <= 1 {
+		_ = level.Error(request.log).Log(logMsg, "unable to find related stateful sets - continuing with single zone")
+		otherStatefulSets = &appsv1.StatefulSetList{Items: []appsv1.StatefulSet{*sts}}
+	}
+
+	// this is the partition of the pod being evicted - for classic zones the partition will be an empty string
+	partition := PdbConfig.PodPartition(pod)
+	matcher := partitionMatcher{
+		same: func(pd *corev1.Pod) bool {
+			return PdbConfig.PodPartition(pd) == partition
+		},
+	}
+
+	// include the partition we are considering
+	if len(partition) > 0 {
+		request.log.SetSpanAndLogTag(logPartition, partition)
+	}
+
+	// include the number of zones we will be considering
+	request.log.SetSpanAndLogTag(logZones, len(otherStatefulSets.Items))
+
+	singleZone := len(otherStatefulSets.Items) == 1
+
+	for _, otherSts := range otherStatefulSets.Items {
+		// exclude our StatefulSet from this iteration unless we only have found a single zone
+		if !singleZone && otherSts.UID == sts.UID {
+			continue
+		}
+
+		// find the number of not ready pods on this StatefulSet which match the given criteria
+		result, err := request.clients.podsNotRunningAndReady(&otherSts, pod, matcher)
+		if err != nil {
+			// TODO - should we allow or deny this ??
```

Comment:
I think the safest thing to do would be to deny - better to delay a node drain than cause a service interruption.